### PR TITLE
Add a new API for listing logs and artifacts from a log area

### PR DESCRIPTION
**Pull request by t-persson**
### Applicable Issues

eiffel-community/etos#238

### Description of the Change

Added a v1alpha API for listing logs and artifacts produced by ETOS. This is to be used to reduce the load of the event repositories when ETOS executes many testruns.

### Alternate Designs

I thought about proxying the files directly from the API and have the client always know how to download the files. I decided against it due to the added complexity and not wanting to deal with memory usage in the logarea service. This is something that may be looked at in a v1 release of the API.

### Possible Drawbacks

This API service will get a lot of traffic instead of the event repository. However I do believe that this solution is more scalable than the Eiffel GraphQL API.

### Sign-off

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.

Signed-off-by: Tobias Persson <tobias.persson@axis.com>

### CODE REVIEW

**Comment by t-persson** [comment]
This change works if the changes in #65 are merged. All CI problems are for python code and I have not changed any python code

**Review by fredjn — CHANGES_REQUESTED** [review]

File: pkg/logarea/v1alpha/logarea.go [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+	handler := &LogAreaHandler{a.logger, a.cfg, a.ctx, a.client, a.regex}
+	router.GET("/v1alpha/selftest/ping", handler.Selftest)
+	router.GET("/v1alpha/logarea/:identifier", handler.GetFileURLs)
+}
+
+// Selftest is a handler to just return 204.
+func (h LogAreaHandler) Selftest(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
+	w.Header().Set("Content-Type", "text/plain; charset=utf-8")
+	w.Header().Set("X-Content-Type-Options", "nosniff")
+	w.WriteHeader(http.StatusNoContent)
```
[the remark this answers is not in the document]
Comment (t-persson) [reply pkg/logarea/v1alpha/logarea.go:95]: It tells clients to not try to infer the content type by parsing the output. It is used for security, but it also helped with another case where the browser borked the API response for me.
I added the header to the other handler as well.

File: pkg/logarea/v1alpha/logarea.go [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+// GetFileURLs is an endpoint for getting file URLs from a log area.
+func (h LogAreaHandler) GetFileURLs(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+	ctx, cancel := context.WithTimeout(r.Context(), time.Second*15)
+	defer cancel()
+	directories := make(Response)
+	identifier := ps.ByName("identifier")
+	// Making it possible for us to correlate logs to a specific connection
+	logger := h.logger.WithField("identifier", identifier)
+
+	response, err := h.client.Get(ctx, fmt.Sprintf("/testrun/%s/suite", identifier), clientv3.WithPrefix())
```
[the remark this answers is not in the document]
Comment (t-persson) [reply pkg/logarea/v1alpha/logarea.go:145]: The only assumption I see is that the `identifier` you pass to the API is the same identifier used for `/testrun`. I believe that this is a fair assumption.

File: test/testconfig/testconfig.go [hunk 3]
```diff
@@ -69,3 +69,8 @@ func (c *cfg) LogFilePath() string {
 func (c *cfg) ETOSNamespace() string {
 	return c.etosNamespace
 }
+
+// DatabaseURI returns the URI to the ETOS database.
+func (c *cfg) DatabaseURI() string {
+	return "etcd-client:2379"
```
[the remark this answers is not in the document]
Comment (t-persson) [reply test/testconfig/testconfig.go:75]: There are no tests using it and we will need to mock ETCD for the tests so the return value of this function does not really matter.

**Review by fredjn — APPROVED** [review]

**Pull request merged**